java/ql/src/Security/CWE/CWE-378/TempDirHijackingVulnerability.ql

```diff
@@ -0,0 +1,90 @@
+/**
+ * @name Temporary Directory Hijacking Vulnerability disclosure
+ * @description Detect temporary directory hijack vulnerability
+ * @kind path-problem
+ * @problem.severity error
+ * @precision very-high
+ * @id java/temp-directory-hijacking
+ */
+
+import java
+import semmle.code.java.dataflow.FlowSources
+import DataFlow::PathGraph
+
+/**
+ * All `java.io.File::createTempFile` methods.
+ */
+private class MethodFileCreateTempFile extends Method {
+  MethodFileCreateTempFile() {
+    this.getDeclaringType() instanceof TypeFile and
+    this.hasName("createTempFile")
+  }
+}
+
+private class MethodFileMkdir extends Method {
+  MethodFileMkdir() {
+    getDeclaringType() instanceof TypeFile and
+    hasName("mkdir")
+    or
+    hasName("mkdirs")
+  }
+}
+
+private class MethodFileDelete extends Method {
+  MethodFileDelete() {
+    getDeclaringType() instanceof TypeFile and
+    hasName("delete")
+  }
+}
+
+predicate isDeleteFileExpr(Expr expr) {
+  exists(MethodAccess ma |
+    expr = ma.getQualifier() and
+    ma.getMethod() instanceof MethodFileDelete
+  )
+}
+
+private class TempDirHijackingToDeleteConfig extends DataFlow::Configuration {
+  TempDirHijackingToDeleteConfig() { this = "TempDirHijackingToDeleteConfig" }
+
+  override predicate isSource(DataFlow::Node source) {
+    exists(MethodAccess ma |
+      ma.getMethod() instanceof MethodFileCreateTempFile and
+      source.asExpr() = ma
+    )
+  }
+
+  override predicate isSink(DataFlow::Node sink) { isDeleteFileExpr(sink.asExpr()) }
+}
+
+private class TempDirHijackingFromDeleteConfig extends DataFlow2::Configuration {
+  TempDirHijackingFromDeleteConfig() { this = "TempDirHijackingFromDeleteConfig" }
+
+  override predicate isSource(DataFlow::Node source) { isDeleteFileExpr(source.asExpr()) }
+
+  override predicate isSink(DataFlow::Node sink) {
+    exists(MethodAccess ma |
+      ma.getMethod() instanceof MethodFileMkdir and
+      ma.getQualifier() = sink.asExpr()
+    )
+  }
+}
+
+predicate isSinkConstrainedByIfCheck(DataFlow2::Node sink) {
+  exists(MethodAccess ma, IfStmt ifStmt |
+    sink.asExpr() = ma.getQualifier() and
+    // Data flow from the return value from `mkdirs` into an 'if' check
```

Comment:
Note you might get use within an `&&`, `||`, `== true` and so on, does this work for those cases?

Reply:
It does not, I'd like to expand this so that it uses `Guard`. I was chatting with a few people about this in the slack channel.

https://ghsecuritylab.slack.com/archives/CQJU6RN49/p1602785245063800

Reply:
This might be a good example to work from: https://github.com/github/codeql/blob/main/java/ql/src/Security/CWE/CWE-022/TaintedPath.ql#L21

Alternatively you could quite closely approximate by just checking the return value has any use -- it's probably a small minority of users that call `mkdir`, store its return value, but then don't properly check it (compared to simply throwing it away altogether)

Reply:
To follow up on how to use `Guard` for this: There are in general not a single recipe for this as there are many ways to mix control flow and data flow in a query.
To cover the linked `TaintedPath` example first: That's using a `BarrierGuard`, which is actually just a convenience wrapper that defines a number of data-flow nodes as barriers/sanitizers (which means that data flow is prohibited from using those nodes); it does this by selecting all those data-flow nodes that are expressions whose corresponding control-flow nodes are guarded by the condition evaluating to false ("guarded" is a technical term with the definition: C evaluates to false guards N if and only if every control-flow path reaching N must have passed through the false-edge leaving C).
BarrierGuards are nice, but doesn't fit the use-case in this query, as the flow path stops at the `mkdir` qualifier. They are however implemented in terms of `Guard`, which allows us to reason about such "guarding" relationships between control-flow nodes.
So what we want is to see whether there's a (problematic) use of `temp` that isn't guarded by `temp.mkdir()` evaluating to true. We can approximate problematic uses by, for instance, all subsequent local uses that are not safe and approximate "safe" by saying that the safe uses are the ones that are used in string concatenations (e.g. for error reporting).
So I'd write this predicate something like this:
```
predicate isSinkConstrainedByIfCheck(DataFlow::Node sink) {
  exists(Guard g, MethodAccess ma, Expr unsafeUse |
    any(TempDirHijackingFromDeleteConfig c).isSink(sink) and
    sink.asExpr() = ma.getQualifier() and
    g = ma and
    DataFlow::localExprFlow(sink.asExpr(), unsafeUse) and
    unsafeUse != sink.asExpr() and
    not safeUse(unsafeUse) and
    not g.controls(unsafeUse.getBasicBlock(), true)
  )
}

predicate safeUse(Expr e) {
  exists(AddExpr concat |
    concat.getType() instanceof TypeString and
    concat.getAnOperand() = e
  )
}
```

Reply:
Btw. "controls" and "guards" are synonymous verbs in this context.